File: Languages/Ruby/Day021.rb
# Languages - Ruby - Day 21
# Foundational Concepts: Basics, syntax, loops, conditionals, functions, and variables.

# Create a function that calculates the Greatest Common Divisor (GCD) of two numbers using the Euclidean Algorithm.


def gcd(a, b)
  # Absolute values of a and b
  a = a.abs
  b = b.abs
  return 0 if a == 0 && b == 0
  
  while b != 0
    a, b = b, a % b
  end
  a
end

# Input validation and interactive output
puts "Enter the first number:"
input1 = gets.chomp
if input1 =~ /^-?\d+$/
  num1 = input1.to_i
  puts "Enter the second number:"
  input2 = gets.chomp
  if input2 =~ /^-?\d+$/
    num2 = input2.to_i
    result = gcd(num1, num2)
    if result == 0
      puts "GCD is undefined for inputs 0 and 0."
    else
      puts "The GCD of #{num1} and #{num2} is #{result}."
    end
  else
    # Error message for invalid input
    puts "Invalid input. Please enter a valid integer."
  end
else
  puts "Invalid input. Please enter a valid integer."
end

File: Languages/Rust/Day021.rs
// Languages - Rust - Day 21
// Foundational Concepts: Basics, syntax, loops, conditionals, functions, and variables.

// Create a function that calculates the Greatest Common Divisor (GCD) of two numbers using the Euclidean Algorithm.


use std::io::{self, Write};

// Function to calculate the Greatest Common Divisor (GCD)
fn gcd(mut a: i32, mut b: i32) -> i32 {
    a = a.abs();
    b = b.abs();

    // GCD is undefined for 0 and 0
    if a == 0 && b == 0 {
        return 0;
    }

    // Euclidean Algorithm
    while b != 0 {
        let temp = b;
        b = a % b;
        a = temp;
    }
    a
}

fn main() {
    // Get user input
    let mut input1 = String::new();
    let mut input2 = String::new();
    // Get the first number
    print!("Enter the first number:\n");
    io::stdout().flush().unwrap();
    io::stdin().read_line(&mut input1).expect("Failed to read line");
    let num1: i32 = match input1.trim().parse() {
        Ok(num) => num,
        Err(_) => {
            println!("Invalid input. Please enter a valid integer.");
            return;
        }
    };
    // Get the second number
    print!("Enter the second number:\n");
    io::stdout().flush().unwrap();
    io::stdin().read_line(&mut input2).expect("Failed to read line");
    let num2: i32 = match input2.trim().parse() {
        Ok(num) => num,
        Err(_) => {
            println!("Invalid input. Please enter a valid integer.");
            return;
        }
    };
    // Calculate the GCD
    let result = gcd(num1, num2);
    if result == 0 {
        println!("GCD is undefined for inputs 0 and 0.");
    } else {
        // Print the result
        println!("The GCD of {} and {} is {}.", num1, num2, result);
    }
}
